<a href="https://colab.research.google.com/github/LucjanSakowicz/data-science-bootcamp/blob/main/09_exercises/141_150_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Pandas

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 141](#1)
* [Ćwiczenie 142](#2)
* [Ćwiczenie 143](#3)
* [Ćwiczenie 144](#4)
* [Ćwiczenie 145](#5)
* [Ćwiczenie 146](#6)
* [Ćwiczenie 147](#7)
* [Ćwiczenie 148](#8)
* [Ćwiczenie 149](#9)
* [Ćwiczenie 150](#10)

### <a name='0'></a>Import biblioteki

In [1]:
import numpy as np
import pandas as pd

np.random.seed(42)
np.__version__

'1.22.4'

### <a name='1'></a> Ćwiczenie 141
Podane są dwa obiekty typu _Series_:

```
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))
```
Połącz te dwa obiekty w jeden obiekt typu _DataFrame_ (jako dwie kolumny) i przypisz do zmiennej _df_. Nadaj nazwy kolumn odpowiednio _col1_ oraz _col2_.


In [2]:
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))

df = pd.concat([s1, s2], axis=1)
df = df.rename(columns={0: 'col1', 1:'col2'}) # lub
df.columns = ['col1', 'col2']

### <a name='2'></a> Ćwiczenie 142
Wytnij z obiektu _df_ wiersze, dla których wartość w kolumnie _col2_ jest pomiędzy 0.0 a 1.0 (włącznie).

In [3]:
df[(df['col2'] <= 1.0) & (df['col2'] >= 0.0)]

#lub
df[df['col2'].between(0.0, 1.0)]

col1      col2
1   0.950714  0.314247
6   0.058084  0.067528
9   0.708073  0.110923
11  0.969910  0.375698
18  0.431945  0.822545

### <a name='3'></a> Ćwiczenie 143
Przypisz nową kolumnę o nazwie _col3_, która przyjmie wartość 1 w momencie gdy w kolumnie _col2_ jest wartość nieujemna i przeciwnie -1.

In [4]:
df['col3'] = np.where(df['col2'] > 0, 1, -1)

#lub

df['col3'] = df['col2'].map(lambda x: 1 if x >= 0 else -1)

df

col1      col2  col3
0   0.374540 -1.012831    -1
1   0.950714  0.314247     1
2   0.731994 -0.908024    -1
3   0.598658 -1.412304    -1
4   0.156019  1.465649     1
5   0.155995 -0.225776    -1
6   0.058084  0.067528     1
7   0.866176 -1.424748    -1
8   0.601115 -0.544383    -1
9   0.708073  0.110923     1
10  0.020584 -1.150994    -1
11  0.969910  0.375698     1
12  0.832443 -0.600639    -1
13  0.212339 -0.291694    -1
14  0.181825 -0.601707    -1
15  0.183405  1.852278     1
16  0.304242 -0.013497    -1
17  0.524756 -1.057711    -1
18  0.431945  0.822545     1
19  0.291229 -1.220844    -1

### <a name='4'></a> Ćwiczenie 144
Przypisz nową kolumnę o nazwie _col4_, która przytnie wartości z kolumny _col2_ do przedziału $[-1.0, 1.0]$. 
Innymi słowy, dla wartości poniżej -1.0 należy wstawić wartość -1.0, dla wartości powyżej 1.0 wartość 1.0.


In [5]:
df['col4'] = df['col2']
df['col4'] = np.where(df['col4'] > 1, 1, df['col4'])
df['col4'] = np.where(df['col4'] < -1, -1, df['col4'])
df

#lub

df['col4'] = df['col2'].clip(-1.0, 1.0)

### <a name='5'></a> Ćwiczenie 145
Znajdź 5 największych wartości dla kolumny _col2_.

In [35]:
df['col2'].nlargest(5)

10    2.143944
1     1.088951
19    0.865755
15    0.852433
5     0.679598
Name: col2, dtype: float64

Znajdź 5 najmniejszych wartości dla kolumny _col2_.

In [37]:
df['col2'].nsmallest(5)

12   -2.025143
3    -1.077745
16   -0.792521
6    -0.730367
4    -0.715304
Name: col2, dtype: float64

### <a name='6'></a> Ćwiczenie 146
Wyznacz skumulowaną sumę dla każdej kolumny.

__Wskazówka:__ Użyj metody _pd.DataFrame.cumsum()_.

In [40]:
df.cumsum()

col1      col2  col3      col4
0    0.867072 -0.471932    -1 -0.471932
1    1.780313  0.617019     0  0.528068
2    2.291655  0.681299     1  0.592348
3    2.793172 -0.396446     0 -0.407652
4    3.591467 -1.111750    -1 -1.122956
5    4.241431 -0.432152     0 -0.443358
6    4.943398 -1.162519    -1 -1.173724
7    5.739190 -0.946060     0 -0.957266
8    6.629196 -0.900488     1 -0.911694
9    6.967191 -1.552089     0 -1.563294
10   7.342774  0.591856     1 -0.563294
11   7.436756  1.225775     2  0.070625
12   8.015036 -0.799368     1 -0.929375
13   8.050978 -0.612914     2 -0.742921
14   8.516576 -1.274700     1 -1.404707
15   9.059221 -0.422267     2 -0.552274
16   9.345762 -1.214788     1 -1.344795
17   9.936595 -1.329524     0 -1.459531
18   9.967095 -0.824537     1 -0.954544
19  10.004444  0.041218     2 -0.088789

### <a name='7'></a> Ćwiczenie 147
Wyznacz medianę dla zmiennej _col2_ (inaczej kwantyl rzędu 0.5).

In [41]:
df['col2'].median()

0.054925929499638274

### <a name='8'></a> Ćwiczenie 148
Wytnij wiersze, dla których zmienna _col2_ przyjmuje wartości większe od 0.0.

In [6]:
df[df['col2'] > 0.0]

#lub

df.query("col2 > 0")

col1      col2  col3      col4
1   0.950714  0.314247     1  0.314247
4   0.156019  1.465649     1  1.000000
6   0.058084  0.067528     1  0.067528
9   0.708073  0.110923     1  0.110923
11  0.969910  0.375698     1  0.375698
15  0.183405  1.852278     1  1.000000
18  0.431945  0.822545     1  0.822545

### <a name='9'></a> Ćwiczenie 149
Wytnij 5 pierwszych wierszy obiektu _df_ i przekształć je do słownika.

In [43]:
df.head(5).to_dict()

{'col1': {0: 0.8670723185801037,
  1: 0.9132405525564713,
  2: 0.5113423988609378,
  3: 0.5015162946871996,
  4: 0.7982951789667752},
 'col2': {0: -0.47193186578943347,
  1: 1.088950596967366,
  2: 0.06428001909546277,
  3: -1.0777447779293061,
  4: -0.7153037092599682},
 'col3': {0: -1, 1: 1, 2: 1, 3: -1, 4: -1},
 'col4': {0: -0.47193186578943347,
  1: 1.0,
  2: 0.06428001909546277,
  3: -1.0,
  4: -0.7153037092599682}}

### <a name='10'></a> Ćwiczenie 150
Wytnij 5 pierwszych wierszy obiektu _df_, przekształć je do formatu Markdown i przypisz do zmiennej _df_markdown_.

In [8]:
df_markdown = df.head(5).to_markdown()

Wydrukuj do konsoli zawartość zmiennej _df_markdown_.

In [9]:
print(df_markdown)

|    |     col1 |      col2 |   col3 |      col4 |
|---:|---------:|----------:|-------:|----------:|
|  0 | 0.37454  | -1.01283  |     -1 | -1        |
|  1 | 0.950714 |  0.314247 |      1 |  0.314247 |
|  2 | 0.731994 | -0.908024 |     -1 | -0.908024 |
|  3 | 0.598658 | -1.4123   |     -1 | -1        |
|  4 | 0.156019 |  1.46565  |      1 |  1        |


Skopiuj wynik uruchomienia powyższej komórki i wklej poniżej:

\# tutaj wpisz rozwiązanie

|    |     col1 |      col2 |   col3 |      col4 |
|---:|---------:|----------:|-------:|----------:|
|  0 | 0.37454  | -1.01283  |     -1 | -1        |
|  1 | 0.950714 |  0.314247 |      1 |  0.314247 |
|  2 | 0.731994 | -0.908024 |     -1 | -0.908024 |
|  3 | 0.598658 | -1.4123   |     -1 | -1        |
|  4 | 0.156019 |  1.46565  |      1 |  1        |
